File: sdk/daml-lf/interpreter/src/test/java/hello-world/Template.java
// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

public abstract class Template(templateId: Ref.TemplateId) {

    // FIXME: use reflection to collect all @arg attributes and build an LfValue
    public arg: LfValue = ???

    // TODO: want following to be static
    public static ContractId<T> create[T <: Template]() {
        return ledger.api.createContract(templateId, arg);
    }

    public Boolean precond() {
        return true;
    }

    public List<Ref.Party> signatories() {
        return List.empty;
    }

    public List<Ref.Party> observers {
        return List.empty;
    }
}

File: sdk/daml-lf/interpreter/src/test/scala/hello-world/src/main/scala/com/digitalasset/daml/lf/Template.scala
// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.digitalasset.daml.lf

import com.digitalasset.daml.lf.ledger.api.{LfValue, LfValueParty}

trait Template {

  def arg: LfValue

  def create[T <: Template, CT <: TemplateId[T]]()(implicit companion: CT): Contract[T, CT] = {
    val contractId = ledger.api.internal.createContract(companion.templateId, arg)

    new Contract[T, CT](contractId)
  }

  def precond: Boolean = {
    true
  }

  def signatories: Seq[LfValueParty]

  def observers: Seq[LfValueParty] = {
    Seq.empty
  }
}
